kernel/mem_domain.c

```diff
@@ -81,7 +81,7 @@ void k_mem_domain_init(struct k_mem_domain *domain, u8_t num_parts,
 
 	key = irq_lock();
 
-	domain->num_partitions = num_parts;
+	domain->num_partitions = 0;
```

Comment:
The cleanup and loop unification looks sane.    But.. I. don't see why this is going to optimize anything.  Nothing here is moving anything the compiler can't already see.  Is the generated code even any different?

Reply:
It is optimization, because:
Before:
Adding partitions is (i=0...N-1)
Sanity for each partition is (j=0...N-1) 

After:
Adding partitions is still (i=0...N-1)
Sanity check for each partition is j=0... i-1

So we end up with half of the loop iterations